src/datascience-ui/ipywidgets/kernel/index.ts

```diff
@@ -0,0 +1,263 @@
+    try {
+        stackOfWidgetsRenderStatusByOutputId.push({ outputId: request.outputId, container: request.element });
+        const output = convertVSCodeOutputToExecutResultOrDisplayData(request);
+
+        // tslint:disable-next-line: no-any
+        const model = output.data['application/vnd.jupyter.widget-view+json'] as any;
+        if (!model) {
+            // tslint:disable-next-line: no-console
+            return console.error('Nothing to render');
+        }
+        // tslint:disable: no-console
+        renderIPyWidget(request.outputId, model, request.element);
+    } catch (ex) {
+        console.error(`Failed to render ipywidget type`, ex);
+        throw ex;
+    }
+}
+export function disposeOutput(e: { outputId: string } | undefined) {
+    if (e) {
+        stackOfWidgetsRenderStatusByOutputId = stackOfWidgetsRenderStatusByOutputId.filter(
+            (item) => !(e.outputId in item)
+        );
+    }
+}
+function renderErrorInLastOutputThatHasNotRendered(message: string) {
+    const possiblyEmptyOutputElement = [...stackOfWidgetsRenderStatusByOutputId]
+        .reverse()
+        .find((item) => !item.success);
+    if (possiblyEmptyOutputElement) {
+        //
+        console.log(message);
+    }
+}
+function renderIPyWidget(
+    outputId: string,
+    model: nbformat.IMimeBundle & { model_id: string; version_major: number },
+    container: HTMLElement
+) {
+    // tslint:disable: no-console
+    // console.error('Got Something to render');
+    if (renderedWidgets.has(outputId)) {
+        return console.error('already rendering');
+    }
+    const output = document.createElement('div');
+    output.className = 'cell-output cell-output';
+    const ele = document.createElement('div');
+    ele.className = 'cell-output-ipywidget-background';
+    container.appendChild(ele);
+    ele.appendChild(output);
+    renderedWidgets.add(outputId);
+    createWidgetView(model, ele)
+        .then((w) => {
+            const disposable = {
+                dispose: () => {
+                    // What if we render the same model in two cells.
+                    renderedWidgets.delete(outputId);
+                    w?.dispose();
+                }
+            };
+            outputDisposables.set(outputId, disposable);
+            outputDisposables2.set(ele, disposable);
+            // Keep track of the fact that we have successfully rendered a widget for this outputId.
+            const statusInfo = stackOfWidgetsRenderStatusByOutputId.find((item) => item.outputId === outputId);
+            if (statusInfo) {
+                statusInfo.success = true;
+            }
+        })
+        .catch((ex) => console.error('Failed to render', ex));
+}
+
+let widgetManagerPromise: Promise<WidgetManager> | undefined;
+async function getWidgetManager(): Promise<WidgetManager> {
+    if (!widgetManagerPromise) {
+        // tslint:disable-next-line: promise-must-complete no-any
```

Comment:
Is promise-must-complete false positive here? Having a bit of trouble tracing it. But generally that's not one to disable.

Reply:
It's a false positive. It's not tracing the resolve handler into the subscriptoin